source/request-response/protocol_adapter.c

```diff
@@ -0,0 +1,428 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/mqtt/private/request-response/protocol_adapter.h>
+
+#include <aws/io/event_loop.h>
+#include <aws/mqtt/private/client_impl_shared.h>
+#include <aws/mqtt/private/request-response/weak_ref.h>
+#include <aws/mqtt/private/v5/mqtt5_client_impl.h>
+#include <aws/mqtt/v5/mqtt5_client.h>
+#include <aws/mqtt/v5/mqtt5_listener.h>
+
+/*
+ * New API contract
+ *
+ * Invariant 1: Subscribe is only called from the RR subscription manager when going from 0 to 1 pending operations
+ * Invariant 2: Unsubscribe is only called from the RR subscription manager when there are 0 pending operations, not
+ *   necessarily on the exact transition to zero though.
+ *
+ * Entries are not tracked with the exception of eventstream impl which needs the stream handles to close.
+ * A subscribe failure should not trigger an unsubscribe, only notify the status callback.
+ * Subscription event callback should be {subscribe_success, subscribe_failure, unsubscribe_success,
+ * unsubscribe_failure}. The sub manager is responsible for calling Unsubscribe on all its entries when shutting down
+ * (before releasing hold of the adapter).
+ *
+ * How do we know not to retry unsubscribe failures because a subscribe came in?  Well, we don't retry failures; let
+ * the manager make that decision.  No retry when the weak ref is zeroed either.  The potential for things to go wrong
+ * is worse than the potential of a subscription "leaking."
+ *
+ * On subscribe failures with zeroed weak ref, trust that an Unsubscribe was sent that will resolve later and let it
+ * decide what to do.
+ */
+
+struct aws_mqtt_protocol_adapter *aws_mqtt_protocol_adapter_new_from_311(
+    struct aws_allocator *allocator,
+    struct aws_mqtt_protocol_adapter_options *options,
+    struct aws_mqtt_client_connection *connection) {
+    (void)allocator;
+    (void)options;
+    (void)connection;
+
+    // TODO
+    return NULL;
+}
+
+/******************************************************************************************************************/
+
+struct aws_mqtt_protocol_adapter_5_impl {
+    struct aws_allocator *allocator;
+    struct aws_mqtt_protocol_adapter base;
+    struct aws_weak_ref *callback_ref;
+    struct aws_mqtt_protocol_adapter_options config;
+
+    struct aws_event_loop *loop;
+    struct aws_mqtt5_client *client;
+    struct aws_mqtt5_listener *listener;
+};
+
+static void s_aws_mqtt_protocol_adapter_5_destroy(void *impl) {
+    struct aws_mqtt_protocol_adapter_5_impl *adapter = impl;
+
+    // all the real cleanup is done in the listener termination callback
+    aws_mqtt5_listener_release(adapter->listener);
```

Comment:
When the last reference is dropped to the protocol adapter, it calls this destructor, which drops its reference to the listener and triggers the listener's cleanup/destruction which then checks its config and calls into `s_protocol_adapter_mqtt5_listener_termination_callback`. This feels like unnecessary complexity. Is there a reason why we aren't executing what's happening in `s_protocol_adapter_mqtt5_listener_termination_callback` in this destroy function after releasing the listener? It'd be easier to follow and remove some aws_mqtt5_listener_config settings from needing to be set during `aws_mqtt_protocol_adapter_new_from_5`.

Reply:
We cannot destroy the listener until we know the client will no longer send us events.